packages/pubsub/src/queue.js

```diff
@@ -0,0 +1,200 @@
+ *         return {
+ *           data: message.data,
+ *           attributes: message.attrs
+ *         };
+ *       })
+ *     };
+ *
+ *     return self.api.Publisher.publish(reqOpts, function(err, resp) {
+ *       if (err) {
+ *         done(err);
+ *         return;
+ *       }
+ *
+ *       done(null, {
+ *         responses: resp.messageIds
+ *       })
+ *     });
+ *   }
+ * });
+ */
+function Queue(options) {
+  this.maxBytes = options.maxBytes || Math.pow(1024, 2) * 5;
+  this.maxMessages = options.maxMessages || 1000;
+  this.maxMilliseconds = options.maxMilliseconds || 1000;
+  this.sendHandler = options.send;
+
+  this.inventory = {
+    queued: new Set(),
+    inFlight: new Set()
+  };
+
+  Object.defineProperty(this.inventory, 'queueBytes', {
+    get: function() {
+      var size = 0;
+
+      this.queued.forEach(function(message) {
+        size += message.data.size;
+      });
+
+      return size;
+    }
+  });
+
+  this.intervalHandle_ = null;
+}
+
+/**
+ * Adds message to the queue.
+ *
+ * @param {*} data - The message data.
+ * @param {function} callback - Callback to be ran once message is sent.
+ *
+ * @example
+ * queue.add(1234, function(err, resp) {});
+ */
+Queue.prototype.add = function(data, callback) {
+  var hasMaxMessages = (this.inventory.queued.size + 1) >= this.maxMessages;
+  var hasMaxBytes = (this.inventory.queueBytes + data.size) >= this.maxBytes;
+
+  if (hasMaxMessages || hasMaxBytes) {
+    this.send(this.getNextMessageBatch());
+  } else if (!this.intervalHandle_) {
+    this.startSendInterval();
+  }
+
+  this.inventory.queued.add({
+    data: data,
+    callback: callback || common.util.noop
+  });
+};
+
+/**
+ * Starts the send interval.
+ *
+ * @example
+ * queue.startSendInterval();
+ */
+Queue.prototype.startSendInterval = function() {
+  if (this.intervalHandle_) {
+    return;
+  }
+
+  var self = this;
+
+  this.intervalHandle_ = setInterval(function() {
+    self.send(self.getNextMessageBatch());
+
+    if (self.inventory.queued.size === 0) {
+      clearInterval(self.intervalHandle_);
+      self.intervalHandle_ = null;
+    }
+  }, this.maxMilliseconds);
+};
+
+/**
+ * Prepares next queue of messages to be sent.
+ *
+ * @return array
+ *
+ * @example
+ * var messages = queue.getNextMessageBatch();
+ * queue.send(messages);
+ */
+Queue.prototype.getNextMessageBatch = function() {
+  var messages = Array.from(this.inventory.queued);
+
+  this.inventory.queued.clear();
+  return messages;
+};
+
+/**
+ * Sends out messages
+ *
+ * @param {array} messages - Array of messages.
+ *
+ * @example
+ * var messages = [{
+ *   data: 123,
+ *   callback: function(err, resp) {}
+ * }];
+ *
+ * queue.send(messages);
+ */
+Queue.prototype.send = function(messages) {
+  if (this.activeRequests_ >= this.maxRequests) {
+    this.scheduleSend();
+    return;
+  }
+
+  var self = this;
+
+  messages.forEach(function(message) {
+    self.inventory.inFlight.add(message);
```

Comment:
I don't think you should retain in-flight messages. It's a waste of memory. The only thing we need to really retain is the callbacks. Can you remove `self.inventory.inFlight` altogether?

I also think it would dramatically simplify the code. `Queue` should only be concerned with messages while they are enqueued. Once you flush the queue, you hand the messages back to `Publisher` via `sendHandler` for both sending the request and dealing with the response... without the callback on lines 190-198. Let `sendHandler` deal with it entirely from that point on.